xbout/region.py

```diff
@@ -0,0 +1,589 @@
+        of this region from the global DataArray.
+
+        Parameters
+        ----------
+        mxg : int
+            Number of guard cells
+        myg : int, optional
+            Number of y-guard cells to include at the corners
+        """
+        ylower = self.ylower_ind
+        if self.connection_lower is not None:
+            ylower -= myg
+        yupper = self.yupper_ind
+        if self.connection_upper is not None:
+            yupper += myg
+        return slice(self.xinner_ind - mxg, self.xinner_ind), slice(ylower, yupper)
+
+    def getOuterGuardsSlices(self, *, mxg, myg=0):
+        """
+        Return x- and y-dimension slices that select mxg guard cells on the outer-x side
+        of this region from the global DataArray.
+
+        Parameters
+        ----------
+        mxg : int
+            Number of guard cells
+        myg : int, optional
+            Number of y-guard cells to include at the corners
+        """
+        ylower = self.ylower_ind
+        if self.connection_lower is not None:
+            ylower -= myg
+        yupper = self.yupper_ind
+        if self.connection_upper is not None:
+            yupper += myg
+        return slice(self.xouter_ind, self.xouter_ind + mxg), slice(ylower, yupper)
+
+    def getLowerGuardsSlices(self, *, myg, mxg=0):
+        """
+        Return x- and y-dimension slices that select myg guard cells on the lower-y side
+        of this region from the global DataArray.
+
+        Parameters
+        ----------
+        myg : int
+            Number of guard cells
+        mxg : int, optional
+            Number of x-guard cells to include at the corners
+        """
+        xinner = self.xinner_ind
+        if self.connection_inner is not None:
+            xinner -= mxg
+        xouter = self.xouter_ind
+        if self.connection_outer is not None:
+            xouter += mxg
+        return slice(xinner, xouter), slice(self.ylower_ind - myg, self.ylower_ind)
+
+    def getUpperGuardsSlices(self, *, myg, mxg=0):
+        """
+        Return x- and y-dimension slices that select myg guard cells on the upper-y side
+        of this region from the global DataArray.
+
+        Parameters
+        ----------
+        myg : int
+            Number of guard cells
+        mxg : int, optional
+            Number of x-guard cells to include at the corners
+        """
+        xinner = self.xinner_ind
+        if self.connection_inner is not None:
+            xinner -= mxg
+        xouter = self.xouter_ind
+        if self.connection_outer is not None:
+            xouter += mxg
+        return slice(xinner, xouter), slice(self.yupper_ind, self.yupper_ind + myg)
+
+
+def _in_range(val, lower, upper):
+    if val < lower:
+        return lower
+    elif val > upper:
+        return upper
+    else:
+        return val
+
+
+def _order_vars(lower, upper):
+    if upper < lower:
+        return upper, lower
+    else:
+        return lower, upper
+
+
+def _get_topology(ds):
```

Comment:
Is this code ever going to be reached for a `slab` topology? Or `s-alpha`?

Reply:
`s-alpha` will reach this and will be labelled as `limiter`, which is a bit more generic. We don't have a specific `slab` geometry at the moment, but if we had one it could actually call the `_create_regions_toroidal()` function, and would be set up as `core`, `sol` or `limiter` (depending on `ixseps1`). The parallel interpolation isn't really needed for a slab though, because the slab is orthogonal so the perpendicular plane is actually an x-z plane.

I think at the moment, we could just call `_create_regions_toroidal` for any BOUT++ simulation because it covers every topology currently possible, but it's conceivable that at some point in the far future there might be a more general topology that couldn't be specified by `ixseps*` and `jyseps*`...

Reply:
`limiter` makes more sense. 

> if we had one it could actually call the _create_regions_toroidal() function, and would be set up as core, sol or limiter (depending on ixseps1).

This would be useful if anyone goes forward and implements e.g. region-aware differencing.

> The parallel interpolation isn't really needed for a slab though, because the slab is orthogonal so the perpendicular plane is actually an x-z plane.

I more just wanted to make sure it behaves gracefully if someone tries to use these methods for a slab geometry.